### PR TITLE
feat: PoC RxSignals

### PR DESCRIPTION
**Pull request by BioPhoton**
- [x] implement
- [x] test
- [x] web docs
- [x] JS docs

---

Often setters are enough. In such cases, we can use state.set({prop}).

If we want to apply behaviour we need to get a stream. This often leads to bloated code with Subjects.

```typescript
@Component({
template: `
<input (input)="searchBtn.next($event.target.value)" /> {{search$ | async}}
<button (click)="submitBtn.next()" >Submit<button>
`
})
class Component {

  _submitBtn = new Sunject<void>();
  _search = new Sunject<string>();
  
  set submitBtn() {
      _submitBtn.next()
  }
  set search(search: string) {
      _search.next(search)
  }
  get search$() {
      return _search.asObservable();
  }
  
  refresh$ = this. search$.pipe(
    switchMap(fetchData)
  )

}
``` 

We could use a Proxy object and a little TS to reduce the boilerplate here:

```typescript
@Component({
template: `
<input (input)="uiActions.searchBtn($event.target.value)" /> {{search$ | async}}
<button (click)="uiActions.submit()" >Submit<button>
`
})
class Component {
  
  uiActions = getActions<{submit: viod, search: string}>()
  
  refresh$ = this. uiActions.search$.pipe(
    switchMap(fetchData)
  )

}
``` 

A config object could even get rid of the mapping in the template:

```typescript
@Component({
template: `
<input (input)="uiActions.search($event)" /> {{search$ | async}}
<button (click)="uiActions.submit()" >Submit<button>
`
})
class Component {
  
  uiActions = getActions<{submit: viod, search: string}>({search: (e) => e.target.value})
  
  refresh$ = this. uiActions.search$.pipe(
    switchMap(fetchData)
  )

}
``` 

If we use the actions only in the class with `RxEffects` or `RxState`  we won't cause any memory leaks.

```typescript
@Component({
template: `
<input (input)="uiActions.search($event.target.value)" /> {{search$ | async}}
<button (click)="uiActions.submit()" >Submit<button>
`
})
class Component {
  
  uiActions = getActions<{submit: viod, search: string}>()
  
  refresh$ = this. uiActions.search$.pipe(
    switchMap(fetchData)
  )

}
``` 

A config object could even get rid of the mapping in the template:

```typescript
@Component({
template: `
<input (input)="uiActions.search($event)" /> {{search$ | async}}
<button (click)="uiActions.submit()" >Submit<button>
`,
providers: [ RxState, RxEffects]
})
class Component {
  
  uiActions = getActions<{submit: viod, search: string}>({search: (e) => e.target.value})
  
  constructor( private effects: RxEffects, private state: RxState<{list: any[]}>) {
    this.state('list', this. uiActions.search$.pipe(switchMap(fetchData)));
    this.effects(this.uiActions.search$, fetchData);
  }

}
``` 

As we have no subscriber after the component is destroyed we don't need to complete the Subjects.

However if we would pass it to another service which has a longer life time we could create a memory leak:

```typescript
@Component({
template: `
<input (input)="uiActions.search($event)" /> {{search$ | async}}
<button (click)="uiActions.submit()" >Submit<button>
`,
providers: [ RxState, RxEffects]
})
class Component {
  
  uiActions = getActions<{submit: viod, search: string}>({search: (e) => e.target.value})
  
  constructor( private globalService: GlobalService) {
    this.globalService.connectSearchTrigger(this.uiActions.search$);
  }

}
```
Here, as the global service lives longer than the component we have subscribers on the subject after the component is destroyed.
For such situations, a hook is needed to get the component destroyed event.

---

**Update:**
Due to typing issues, I had to refactor to an architecture where we have a wrapper scope for the action create function.
Incredible BIG THANKS to @ddprrt for the support here. 

The current solution provides a service and a factory function. 

The service is hooked into Angular's life-cycles and so it cleans up on destruction.
The factory function returns a pair of functions, create and destroy that can be used manually.

Service:
```typescript
export class RxActionFactory<T extends Actions> implements OnDestroy {
  private readonly subjects: SubjectMap<T> = {} as SubjectMap<T>;
  create<U extends ActionTransforms<T> = {}>(transforms?: U): RxActions<T, U> {
    return new Proxy(
      {} as RxActions<T, U>,
      actionProxyHandler(this.subjects, transforms)
    ) as RxActions<T, U>;
  }

  destroy() {
    for (let subjectsKey in this.subjects) {
      this.subjects[subjectsKey].complete();
    }
  }

  ngOnDestroy() {
    this.destroy();
  }
}
```

Service Usage:
```typescript
interface StateActions {
  refreshList: viod,
}

@Injectable({providedIn: 'root'})
class GlobalState implements OnDestroy {
    _fac = new RxActionFactory<StateActions >();
    uiActions = this._fac.create()
    
    ngOnDestroy() {
        this._fac.destroy();
    }
```

Component Usage:
```typescript
interface UIActions {
  submit: viod, 
  search: string
}

@Component({
template: `
<input (input)="uiActions.search($event)" /> {{search$ | async}}
<button (click)="uiActions.submit()" >Submit<button>
`,
providers: [ RxState, RxEffects, RxActionFactory]
})
class Component {
  
  uiActions = this.actionFactory.create({
    search: (e: Event | string | number) => e.target.value !== undefined ? e.target.value + '' : e + ''
  })
  
  constructor( 
     private globalService: GlobalService,
     private actionFactory: RxActionFactory<UIActions>
  ) {
    this.globalService.connectSearchTrigger(this.uiActions.search$);
  }

}
```

The factory function is maybe irrelevant for Angular but still worth looking at it:

```typescript
export function rxActionCreator<T extends {}>() {
  const subjects = {} as SubjectMap<T>;
  return {
    create,
    destroy: (): void => {
      for (let subjectsKey in subjects) {
        subjects[subjectsKey].complete();
      }
    },
  };

  function create<U extends ActionTransforms<T> = {}>(
    transforms?: U
  ): RxActions<T, U> {
    return new Proxy(
      {} as RxActions<T, U>,
      actionProxyHandler(subjects, transforms)
    ) as RxActions<T, U>;
  }
}
```

Usage:
```typescript
type UIActions = {
  search: string;
  check: number;
};

const {create, destroy} = rxActionCreator<UIActions>();
const actions = create({
  search: (v: number) => 'string',
});
create.search(4);
```

---

**Update:**

I removed the function because we can also do `new RxActionsFactory<any>().create()`.
Additionally, docus are now present as well as tests.

**Missin features:**
- Angular Error handler

**Missing tests:**
- subscribe to a subject that is not emitted on
- emit on a property without subscriber
- error in transform

--- 

The only flaw we could maybe see here is the wrapper e.g. new RxActionsFactory().create().prop

For now I have no clue how we could get rid of it and expose the setter and observables directly e.g. new RxActionsFactory().prop.
  One problem is typescript, the other Proxies, classes and typescript :). Maybe problems to solve in a later step...

Related issues:
#423, #1013

### CODE REVIEW

**Comment by nartc** [comment]
```ts
type InstanceOrType<T> = T extends abstract new (...args: unknown[]) => infer R
  ? R
  : T;

type UiInteractions<T extends { [x: string]: any }> = {
  [K in keyof T]: (arg: InstanceOrType<T[K]>) => void;
} & {
  [K in Extract<keyof T, string> as `${K}$`]: Observable<T[K]>;
};

export function getActions<T extends object>(): UiInteractions<T> {
  const subject = new Subject<Partial<T>>();

  const handler: ProxyHandler<UiInteractions<T>> = {
    get(target, property) {
      const propName = property.toString().slice(0, -1);

      if (property.toString().split('').pop() === '$') {
        return subject.pipe(map((t) => t[propName as keyof T]));
      }

      return (args: T[keyof T]) => {
        subject.next({ [property]: args } as Partial<T>);
      };
    },
    set(target, property, vale) {
      throw new Error('No setters available. To emit call the property name.');
    },
  };

  return new Proxy({} as UiInteractions<T>, handler);
}

```

This is a slightly updated typings version as the original has a couple of issues:
- `handler` needs to be typed to `ProxyHandler<UiInteractions<T>>`
- `${K}$` in `UiInteractions` needs to be relaxed to `Observable<>` so that it is compatible with `hold()` or `hold()` needs to be tighten its parameters to `Subscribable<>`

**Comment by BioPhoton** [comment]
Hi @nartc ,

What do you say to the config object `{search: (e) => e.target.value}`?

It would centralize all customizations needed to consume the UI.

**Comment by nartc** [comment]
@BioPhoton basically like `projections` for the Subject right? I think it is nice to have but it is not too big of a deal if we cannot have it. Typings would be tricky I think.

**Comment by BioPhoton** [comment]
@nartc you can have a look at the implementation and typing of the transform.

**Comment by ddprrt** [comment]
You might want to change the return value to `Observable<InstanceOrType<T[K]>>` or you'll end up with constructor functions instead of actual objects.

```diff
type UiInteractions<T extends { [x: string]: any }> = {
  [K in keyof T]: (arg: InstanceOrType<T[K]>) => void;
} & {
-  [K in Extract<keyof T, string> as `${K}$`]: Observable<T[K]>;
+  [K in Extract<keyof T, string> as `${K}$`]: Observable<InstanceOrType<T[K]>>;
};
```

**Comment by nartc** [comment]
```ts
@Component({
template: `
<input (input)="uiActions.searchBtn($event)" /> {{search$ | async}}
<button (click)="uiActions.submit()" >Submit<button>
`
})
class Component {
  
  uiActions = getActions<{submit: viod, search: string}>({search: (e) => e.target.value})
  
  refresh$ = this. uiActions.search$.pipe(
    switchMap(fetchData)
  )

}
```

This example would fail template type check. `uiActions.search()` expects a `string` (based on the current implementation) but `$event` here is of type `Event`.

**Comment by BioPhoton** [comment]
Hi @ddprrt! 

Thx for your support!

We think about dropping a feature because of complexity but I would really love to see that shipped.

```typescript

function coerceString(e: Event | string): string {
  if(typeof e === 'string') {
    return e;
  }
  return e.target.value;
}

const uiActions = getActions<{search: string}>({search: coerceString});

// in template
uiActions.search($event); // {target:{value: '42'}}
// in class
uiActions.search('42');
```

Any ideas how I could approach this or if it is even a good idea?

**Comment by BioPhoton** [comment]
@wizardnet972  I realized you commented on a issue on that too... 
Please have a look.

File: libs/state/actions/README.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
+# @rx-angular/cdk/actions
+
+[![npm](https://img.shields.io/npm/v/%40rx-angular%2Fcdk.svg)](https://www.npmjs.com/package/%40rx-angular%2Fcdk)
+![rx-angular CI](https://github.com/rx-angular/rx-angular/workflows/rx-angular%20CI/badge.svg?branch=master)
+[![Coverage Status](https://raw.githubusercontent.com/rx-angular/rx-angular/github-pages/docs/test-coverage/cdk/jest-coverage-badge.svg)](https://rx-angular.github.io/rx-angular/test-coverage/cdk/lcov-report/index.html)
+
+## This package provides a quick way to create an action challel and some configuration options.
```
Comment (BioPhoton) [inline libs/state/actions/README.md:7]:
```suggestion
## This package provides a quick way to create an action channel and some configuration options. 
```

File: libs/state/actions/README.md [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+```bash
+npm install --save @rx-angular/cdk
+# or
+yarn add @rx-angular/cdk
+```
+
+## Documentation
+
+- [Actions](https://github.com/rx-angular/rx-angular/tree/master/libs/cdk/actions/docs/Readme.md)
```
Comment (BioPhoton) [inline libs/state/actions/README.md:31]:
```suggestion
# @rx-angular/state/actions

[![npm](https://img.shields.io/npm/v/%40rx-angular%2Fstate.svg)](https://www.npmjs.com/package/%40rx-angular%2Fstate)
![rx-angular CI](https://github.com/rx-angular/rx-angular/workflows/rx-angular%20CI/badge.svg?branch=master)
[![Coverage Status](https://raw.githubusercontent.com/rx-angular/rx-angular/github-pages/docs/test-coverage/state/jest-coverage-badge.svg)](https://rx-angular.github.io/rx-angular/test-coverage/state/lcov-report/index.html)

## This package provides a quick way to create an action channel and some configuration options. 


## Key features

 - ✅ Fully Typed
 - ✅ No-Boilerplate
 - ✅ Configurable transformations to have lines in the template
 - ✅ Minimal memory footprint through a Proxy object and lazy initialization
 
## Demos:

- [⚡ GitHub](https://github.com/BioPhoton/rx-angular-state-coalescing)

## Install

```bash
npm install --save @rx-angular/state
# or
yarn add @rx-angular/state
```

## Documentation

- [Actions](https://github.com/rx-angular/rx-angular/tree/master/libs/state/actions/docs/Readme.md)
```

File: libs/state/actions/docs/Readme.md [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+// or
+yarn add @rx-angular/state
+```
+
+## Usage
+
+As actions are used in global state management as well as in UI event composition we will have a look on both ends.
+
+
+There are 2 common ways to create an action stream and its emission:
```
Comment (edbzn) [inline libs/state/actions/docs/Readme.md:120]: I find this section a bit confusing, we expect the usage of the library and not alternatives without.

File: libs/state/actions/docs/Readme.md [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+    );
+  }
+
+}
+```
+
+### Usage for template events
+
+In components/templates we can use the optional `transforms` param to reduce repetativ code in the template.
+In addition we can use it as a shorthand in the template and directly connect to action dispatching in the calss.
```
Comment (edbzn) [inline libs/state/actions/docs/Readme.md:267]:
```suggestion
In addition, we can use it as a shorthand in the template and directly connect to action dispatching in the class.
```

File: libs/state/actions/docs/Readme.md [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+      )
+    );
+  }
+
+}
+```
+
+### Usage for template events
+
+In components/templates we can use the optional `transforms` param to reduce repetativ code in the template.
```
Comment (edbzn) [inline libs/state/actions/docs/Readme.md:266]:
```suggestion
In components/templates we can use the optional `transforms` param to reduce repetitive code in the template.
```

**Review by edbzn — COMMENTED** [review]
Love this package, great work :100:

File: libs/state/actions/docs/Readme.md [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+
+**RxA - Actions**
+```typescript
+interface Commands {
+  refreshUser: string | number;
+  refreshList: string | number;
+  refreshGenres: string | number;
+}
+
+ commands = getActions<>()<Commands>();
```
Comment (BioPhoton) [inline libs/state/actions/docs/Readme.md:203]:
```suggestion
 commands = getActions<Commands>();
```

[the remark this answers is not in the document]
Comment (BioPhoton) [reply libs/state/actions/docs/Readme.md]: done

File: libs/state/actions/docs/Readme.md [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+
+```typescript
+interface Commands {
+  refreshUser: string | number;
+  refreshList: string | number;
+  refreshGenres: string | number;
+}
+```
+
+Next we can use the typing the create the action object:
```
Comment (eneajaho) [inline libs/state/actions/docs/Readme.md:127]:
```suggestion
Next we can use the typing to create the action object:
```

File: libs/state/actions/src/lib/actions.factory.ts [hunk 8]
```diff
@@ -0,0 +1,84 @@
+import { Actions, ActionTransforms, RxActions } from './types';
+import { Injectable, OnDestroy } from '@angular/core';
+import { actionProxyHandler } from './proxy';
+import { Subject } from 'rxjs';
+
+type SubjectMap<T> = { [K in keyof T]: Subject<T[K]> }
+
+/**
+ * This class to creates RxActions bound to Angular's DI life-cycles. This prevents memorie leaks and optionally makes the instance reusable across the app.
```
Comment (eneajaho) [inline libs/state/actions/src/lib/actions.factory.ts:9]:
```suggestion
 * This class creates RxActions bound to Angular's DI life-cycles. This prevents memory leaks and optionally makes the instance reusable across the app.
```

File: libs/state/actions/src/lib/actions.factory.ts [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+ * The main function here is called `create`, optionally you can also call `destroy` to complete all action channels. 
+ * If the instanciator gets destroyed also the actions get destroiyed automatically.
```
Comment (eneajaho) [inline libs/state/actions/src/lib/actions.factory.ts:11]:
```suggestion
 * If the instantiator gets destroyed also the actions get destroyed automatically.
```

**Comment by wizardnet972** [comment]
@BioPhoton I think you also should add `next$`, `error$`, `complete$` observables to hook when action is in those states in order to reduce or reuse the flow:

```
 ui = this.factory.create(...);
 
 readnoly create$ = this.ui.create$.pipe(...);
 
 readonly createNext$ = this.ui.createNext$.pipe(...)
```

<action>
<action>$
<action>Next$
<action>Error$
<action>Complete$

**Comment by wizardnet972** [comment]
Hey @BioPhoton , 

Why when I add `startWith` to `load$` pipe -  `exhaustMap` does not work?
`endWith` also does not work.

I made an example application on `stackblitz` **with the last update files from your POC**: https://stackblitz.com/edit/angular-ivy-z6ca6k?file=src%2Fapp%2Ffoo.service.ts

You can see the state as json in the output.
 
```
export interface UiActions {
  load: void;
}

@Injectable({ providedIn: 'root' })
export class AppFacade extends RxState<{ items: any[]; pending: boolean }> {
  ui = this.factory.create();

  load$ = this.ui.load$.pipe(
    exhaustMap(() =>
      of({ items: [{ id: 1 }, { id: 2 }] }).pipe(delay(4 * 1000))
    ),
    startWith({ pending: true }),
    endWith({ pending: false })
  );

  constructor(private factory: RxActionFactory<UiActions>) {
    super();
    this.connect(this.load$);
  }
}
```

**Comment by wizardnet972** [comment]
How can I tell when the action is complete the operators pipeline?

I only dispatch an action: `ui.create(...);` but from wherever I did the dispatch I also want to know when it's complete the command pipeline.

How I do it using this version of rx-actions?

I suggest to return the subject itself to continue the pipeline. `ui.create(...).pipe(...)`

**Comment by wizardnet972** [comment]
@BioPhoton can you implement `tapResponse` function from `ngrx/component-store`?

https://github.com/ngrx/platform/blob/master/modules/component-store/src/tap-response.ts

Because right now, the actions lead to error-boilerplate and the file is too long:

see [stackblitz example I made](https://stackblitz.com/edit/angular-ivy-rve5gk?file=src%2Fapp%2Fapp-state.service.ts,src%2Fapp%2Fapp.component.ts)
```
 readonly action1$ = this.ui.action1$.pipe(
    exhaustMap(() =>
      this.api.saveData().pipe(
        switchMap(() =>
          this.api.saveData2().pipe(
            tap({
              next: () => {
                console.log('NOTIFICATION: save');
              },
            })
          )
        ),
        catchError((e) => {
          console.log('ERROR!!');
          return of(e);
        })
      )
    )
  );

```

with tapResponse:
```
readonly action1$ = this.ui.action1$.pipe(
    exhaustMap(() =>
      this.api.saveData().pipe(
        switchMap(() => this.api.saveData2()),
        tapResponse(
          () => console.log('NOTIFICATION: save'),
          () => console.log('ERROR!!')
        )
      )
    )
  );
```

**Comment by wizardnet972** [comment]
Another issue I have is to run another action and wait for a response:

[You can see it in stackblitz example I made](https://stackblitz.com/edit/angular-ivy-4ytug9?file=src%2Fapp%2Fapp-state.service.ts,src%2Fapp%2Fapi.service.ts)

From action1 I want to trigger action2 and after that return to action1 after the pipeline of action2 is done.

```
 readonly action1$ = this.ui.action1$.pipe(
    exhaustMap(() =>
      this.api.saveData().pipe(
        switchMap(() =>
          this.api.saveData2().pipe(
            tap({
              next: () => {
                console.log('NOTIFICATION FROM ACTION1: save');

                // this.ui.action2() after that log into the console.
                /**
                 * this.ui.action2().pipe(tap(() => console.log('after action 2')))
                 */
              },
            })
          )
        ),
        catchError((e) => {
          console.log('ERROR!!');
          return of(e);
        })
      )
    )
  );
```

**Comment by wizardnet972** [comment]
One more thing: when error is throw the entire action is stopped (can't emit the action again) unless using `catchError` in the pipeline:

```
 saveData() {
   throw new Error('bad');
   ...
 }
 ...
 readonly action1$ = this.ui.action1$.pipe(
    exhaustMap(() =>
      this.api.saveData().pipe(
        switchMap(() => ...
     ....   
```

**Comment by jon9090** [comment]
When you plan to publish this PR?

**Comment by yjaaidi** [comment]
@BioPhoton as a follow-up to our conversation about RxSignals, I still didn't dive in depth in the PR but concerning the transformers typing issue, I think I've got something... except if I didn't get the problem 😅

Actually, there is an issue with TypeScript generics and multiple parameters... it's all or nothing. Either everything is infered, or we have to hint all the types: https://github.com/microsoft/TypeScript/issues/10571
That's why I think that there are currently only two solutions and I hope that I am wrong 😄.

A.1. - Splitting the signal param types which are given explicitly and the transformers which are infered through class + method:
```ts
new Signals<{search: string}>>
  .create({
    search(value: number | string) {
      return value.toString();
    }
  })
```
A.2. - or builder pattern
 ```ts
 createSignals<{search: string}>()
   .withTransformers({
     search(value number | string) {
       return value.toString()
     }
   })
 ```

B - Explicit typing:

```ts
createSignals<
    { search: string; reset: void },
    { search: number | string } // transformers params
>({
    search(value) { // type is infered
      return value.toString();
    },
});
```

Personally, I am more in favour of A.2 as it's the less verbose and more explicit in the most common case where we don't need transformers:
```ts
createSignals<{search: string}>();
```

[open in typescript playground](https://www.typescriptlang.org/play?#code/PQKhCgAIUghBXAlgGwCYFMBOkAOBDAFwKwDtI8cdMB7PAYwAsA6KEYcEvAW3QGd866SAEFIAbyiQ61ErwLk6BRDN6QAvFMzpC6YYuWyAPBMinIvbZkYAucwUyISAcwDck01osFbAN2qJUN1MAXwA+AAoASiYAd0QCBgAVTDxZADNqTB5MXnCTMws8KwZwnzxkeHRbEnguACMsSAAfOwdnSNs5NqdxdzMtAnhMMjKK9CYCagBle0cnKKCzYIAaSWDIt0l6JRUmQuLwgEYNyWBgMwuAPQB+Lf1d-cZwgG0AXRPTM4uzG7ud2T2lieAHIMtRgR9IF9vpBfqZtgZeICiowACSLaHfOEKf5IzzoAgLU7nGHYhG7fEEdHEmGw26SNLwEj3Mh0LQ6PS4wyJSDoAAexBIqFU1DqACt0IoIpFehcBkMyPkLnEEslUrwMlksLxuQAlYQAOSmADEAPK6gCyAFFdVNeQL0ELVGr0plsjqAApFJTlbmhf3hPoXezqzXu2yJfVGs2Wm1TIOQGVK77y4aQGrIZDkVSpACe2ZELIAIoh+IRGNq-ZAAGQJi6cgymuoWTBlOrIPhV6uy2kXZ4AaUgjkgAGt0LnqGlIJHDSbzdbba9bIHezDRpVbF6UjxiDk9bOYwupgPXqFngAGV5174ytShSB+AKLXvBZ9LVYXV9rcCSAi5nBCA2MglmWBAVnuPL8oKwqQKKEpSuoPamAOQ5kGOE5TokS6QKU5QbtOJ63vej6BGsmymH+AGFv8TYtm2HY6pBDpOrB4qSgQ94aMmKHDlaAopIoMzdFWeCqAABqiYj9sEqJidhtFYPR6DcieoSLF+v7-kILoam6lZMdBIpsQhXF9DxaHjpO07YbhYy2HmREPv4pEhORqG7mk9BCAprZ4O2ymJPeya8PAzZsogDRRL4znqT+FFaZAfEhoJszONy9qGax8EcYhSUCQQhjoVZiTLK0cxqeAwQ-qAEDQIlfI4MgiB0PEkCUXwLDQOwnA8GWghwD20iyPI5KyIhbLaMQQFGH0YjmECDCdKlriQJS0UBJAKyzfNKKLemtQNNgLRdHMm1QucIaulqOS4EU3C8JIETBQttmVEmCapiMeHjJMQlzESn4fptJx-IiyIHMcGIklitzwiySKPCUbyQpiVywziYOI+EoLUOCKPQ2joMPAt1KfATPzo6NeJ8ASAPnbSZLw0wlKk-TpL0qYjLMv8miTbo8OGH0wgAMKJAAkqaRoZY6MFwexQOmDO0bznG0ssTpYaVluPrIIYIvi5LUz+o94SXbp128LY3GDsORVTtriC+vrEtGqENnrlU05RnOsa2oR6j3s7hsnup71ygSCr7ZmBZ5gW00gfgYEMJWQeuzWCbTT5Sk6qnRs1kh3zmaOlmYd7h5xu730RmXKt+-27wB05T4Jl+IRxW1CXx6WifgYxauy8ZOWmX2NsWRh1nLh7Eb+3eTcuZtbntdRjbNopfkMelUEy0Z2WcQXRd5dsf1pYFBYSVJMlybYWfrwFqnqYvCUa3pEH9zv7F79bqHF+PWH2SQuYH7t0cB5LykAb7+SrMFUKvBwqRQ6HPWKmkqKHxSsJAy28sof1yvxbYhUS7TlKidZwFUqrgBqqwSAAB1TII4ijUCZKgSADRNRCDkAdJqzhIDwBwDIOAnU2AcG4HwAQQhhaDRUCNeG412RTQFttRGS1uguFWjTbwc9NoK3EDtYok8q77XqI0Y6y1HIkU2ibZ6u1XroDDimCOaYPYTGmMtOmIQgbrDclTcGTxIY0nZkTAEWNkYqK+IAXg3AClO34mGASEYvRxnjKGDNKZM0RuiemECOyGGIU4UIUTCZw1xMzNRdNUYUxiUUrwaSvgZOUiRXJZMknty5iyXmHJ5EXAwSxOWigtFKx9keN+04UhXXdIYB2vpAr1MgBEM2mscjVwPLXKYtiPD2LIBmLMolyAALjsWbu5Zk4QXvLWGEmdV6+X8oxY5BcR7fztl7RZvspg2WvKYKekAtzCN3IxGuTzVIXivL2UxMUW6xXilRLuoFe6b2YgPXeiEv62wIX-HC7zEgz2IiC1yyDAIshqX3LeXTB6fzMqPeqyUCDHycCJcSklpKyXkucpSKl64VWxeC7SwzzajM6XC7Bw8zBF3uSi6x-9czAubuy9yWBPL9RqVAvoIUwoOHgeteeVUOXkvylSmFmVulDy1Xg4VRDlqkPAEAA)

```ts
/**
 * Builder pattern approach.
 */
namespace A {
  const actions = createActions<{
    search: string;
    reset: void;
  }>().withTransformers({
    search(value: number | string): string {
      return value.toString();
    },
  });

  actions.search(1);
  //       ^?
  actions.search([]);
  //       ^?
  actions.search('foo');
  //       ^?
  actions.search$;
  //       ^?
  actions.reset();
  //       ^?
  actions.reset$;
  //       ^?

  function createActions<T extends object>() {
    return {
      withTransformers<TRANSFORMERS extends Transformers<Partial<T>>>(
        transformers: TRANSFORMERS
      ) {
        return null as any as ActionDispatchers<T> &
          ActionObservables<T> & {
            [K in keyof TRANSFORMERS]: (
              value: Parameters<TRANSFORMERS[K]>[0]
            ) => void;
          };
      },
    };
  }

  type ActionDispatchers<T extends object> = {
    [K in keyof T]: (value: T[K]) => void;
  };

  type ActionObservables<T extends object> = {
    [K in ExtractString<T> as `${K}$`]: Observable<T[K]>;
  };

  type Transformers<T extends object> = {
    [K in keyof T]: (value: any) => void;
  };

  interface Observable<T> {
    subscribe(): void;
  }

  type ExtractString<T extends object> = Extract<keyof T, string>;
}

/**
 * Explicit types.
 */
namespace B {
  const actions = createActions<
    { search: string; reset: void },
    { search: number | string } // transformers params
  >({
    search(value) {
      return value.toString();
    },
  });

  actions.search(1);
  //       ^?
  actions.search([]);
  //       ^?
  actions.search('foo');
  //       ^?
  actions.search$;
  //       ^?
  actions.reset();
  //       ^?
  actions.reset$;
  //       ^?

  function createActions<
    ACTIONS extends object,
    TRANSFORMERS extends Transformers<Partial<ACTIONS>>
  >(transformers: {
    [K in keyof Partial<ACTIONS>]: (value: TRANSFORMERS[K]) => ACTIONS[K];
  }) {
    return null as any as ActionDispatchers<ACTIONS> &
      ActionObservables<ACTIONS> & {
        [K in keyof TRANSFORMERS]: (value: TRANSFORMERS[K]) => void;
      };
  }

  type ActionDispatchers<T extends object> = {
    [K in keyof T]: (value: T[K]) => void;
  };

  type ActionObservables<T extends object> = {
    [K in ExtractString<T> as `${K}$`]: Observable<T[K]>;
  };

  type Transformers<T extends object> = {
    [K in keyof T]: any;
  };

  interface Observable<T> {
    subscribe(): void;
  }

  type ExtractString<T extends object> = Extract<keyof T, string>;
}

/**
 * Workaround before stumbling upon B.
 */
namespace C {
  const actions = createActions<
    { search: string; reset: void },
    { search: (value: number | string) => void }
  >({
    search(value) {
      return value.toString();
    },
  });

  actions.search(1);
  //       ^?
  actions.search([]); // 💥
  //       ^?
  actions.search('foo');
  //       ^?
  actions.search$; // Observable<string>
  //       ^?
  actions.reset();
  //       ^?
  actions.reset$; // Observable<void>
  //       ^?

  function createActions<
    T extends object,
    TRANSFORMERS extends Transformers<Partial<T>>
  >(transformers: TRANSFORMERS) {
    return null as any as ActionDispatchers<T> &
      ActionObservables<T> & {
        [K in keyof TRANSFORMERS]: (
          value: Parameters<TRANSFORMERS[K]>[0]
        ) => void;
      };
  }
  type ActionDispatchers<T extends object> = {
    [K in keyof T]: (value: T[K]) => void;
  };
  type ActionObservables<T extends object> = {
    [K in ExtractString<T> as `${K}$`]: Observable<T[K]>;
  };
  type Transformers<T extends object> = {
    [K in keyof T]: (value: any) => void;
  };
  interface Observable<T> {
    subscribe(): void;
  }
  type ExtractString<T extends object> = Extract<keyof T, string>;
}
```

**Comment by BioPhoton** [comment]
fixed #1376

**Review by hoebbelsB — APPROVED** [review]
PR looks amazing, let's ship it like this!

**Pull request merged**